icu4c/source/i18n/erarules.cpp

```diff
@@ -0,0 +1,313 @@
+    return startDate != 0;
+}
+
+static UBool isValidRuleStartDate(int32_t year, int32_t month, int32_t day) {
+    return year >= MIN_ENCODED_START_YEAR && year <= MAX_ENCODED_START_YEAR
+            && month >= 1 && month <= 12 && day >=1 && day <= 31;
+}
+
+/**
+ * Encode year/month/date to a single integer.
+ * year is high 16 bits (-32768 to 32767), month is
+ * next 8 bits and day of month is last 8 bits.
+ *
+ * @param year  year
+ * @param month month (1-base)
+ * @param day   day of month
+ * @return  an encoded date.
+ */
+static int32_t encodeDate(int32_t year, int32_t month, int32_t day) {
+    return year << 16 | month << 8 | day;
+}
+
+static void decodeDate(int32_t encodedDate, int32_t (&fields)[3]) {
+    if (encodedDate == MIN_ENCODED_START) {
+        fields[0] = MIN_INT32;
+        fields[1] = 1;
+        fields[2] = 1;
+    } else {
+        fields[0] = (encodedDate & YEAR_MASK) >> 16;
+        fields[1] = (encodedDate & MONTH_MASK) >> 8;
+        fields[2] = encodedDate & DAY_MASK;
+    }
+}
+
+/**
+ * Compare an encoded date with another date specified by year/month/day.
+ * @param encoded   An encoded date
+ * @param year      Year of another date
+ * @param month     Month of another date
+ * @param day       Day of another date
+ * @return -1 when encoded date is earlier, 0 when two dates are same,
+ *          and 1 when encoded date is later.
+ */
+static int32_t compareEncodedDateWithYMD(int encoded, int year, int month, int day) {
+    if (year < MIN_ENCODED_START_YEAR) {
+        if (encoded == MIN_ENCODED_START) {
+            if (year > MIN_INT32 || month > 1 || day > 1) {
+                return -1;
+            }
+            return 0;
+        } else {
+            return 1;
+        }
+    } else if (year > MAX_ENCODED_START_YEAR) {
+        return -1;
+    } else {
+        int tmp = encodeDate(year, month, day);
+        if (encoded < tmp) {
+            return -1;
+        } else if (encoded == tmp) {
+            return 0;
+        } else {
+            return 1;
+        }
+    }
+}
+
+EraRules::EraRules(int32_t *startDates, int32_t numEras)
+    : startDates(startDates), numEras(numEras) {
+    initCurrentEra();
+}
+
+EraRules::~EraRules() {
+    uprv_free(startDates);
+}
+
+EraRules* EraRules::createInstance(const char *calType, UBool includeTentativeEra, UErrorCode& status) {
+    if(U_FAILURE(status)) {
+        return nullptr;
+    }
+    LocalUResourceBundlePointer rb(ures_openDirect(nullptr, "supplementalData", &status));
+    ures_getByKey(rb.getAlias(), "calendarData", rb.getAlias(), &status);
+    ures_getByKey(rb.getAlias(), calType, rb.getAlias(), &status);
+    ures_getByKey(rb.getAlias(), "eras", rb.getAlias(), &status);
+
+    if (U_FAILURE(status)) {
+        return nullptr;
+    }
+
+    int32_t numEras = ures_getSize(rb.getAlias());
+    int32_t firstTentativeIdx = MAX_INT32;
+
+    int32_t *startDates = (int32_t*)uprv_malloc(numEras * sizeof(int32_t));
```

Comment:
nit: could probably use LocalPointer instead of uprv_malloc.  but I don't think this is necessary for now. I can change this later when adding UWP env support.

Reply:
Thanks